File: 백준/Silver/1303. 전쟁 － 전투/전쟁 － 전투.py
import sys
sys.setrecursionlimit(10 ** 6)

M, N = map(int, input().split())

# W, B 경우 나눠서 탐색 후 계산 하고 각각의 배열에 넣고 합 추가

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

arr = [list(input().strip()) for _ in range(N)]


visited = [[0 for _ in range(M)] for _ in range(N)]

w_list = []
b_list = []

def dfs(x, y, color):
    visited[x][y] = 1
    cnt = 1
    for k in range(4):
        nx, ny = x + dx[k], y + dy[k]
        if 0 <= nx < N and 0 <= ny < M:
            if visited[nx][ny] == 0 and arr[nx][ny] == color:
                cnt += dfs(nx, ny, color)
    return cnt
        

for i in range(N):
    for j in range(M):
        if visited[i][j] == 0:
            if arr[i][j] == 'W':
                w_list.append(dfs(i, j, 'W'))
            else :
                b_list.append(dfs(i, j, 'B'))

print(sum(x * x for x in w_list), sum(x * x for x in b_list))

File: 백준/Silver/1303. 전쟁 － 전투/전쟁 － 전투.js
const fs = require("fs");
const path = require("path");
const input = fs
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : path.join(__dirname, "input.txt"))
  .toString()
  .trim()
  .split("\n");

const solution = (input) => {
  const [N, M] = input[0].split(" ").map(Number); // N=가로(M), M=세로(N) 문제에 따라 다름
  // 백준 1303은 "N M" = 가로 N, 세로 M
  // 네 JS는 [N, M]을 행/열로 쓰고 있으니 아래처럼 맞춰주는게 안전:
  const cols = N; 
  const rows = M;

  const arr = Array.from({ length: rows }, (_, i) => input[i + 1].split(""));
  const visited = Array.from({ length: rows }, () => Array(cols).fill(false));

  const dx = [-1, 1, 0, 0];
  const dy = [0, 0, -1, 1];

  const bfs = (sx, sy, color) => {
    let head = 0;
    const q = [[sx, sy]];
    visited[sx][sy] = true;
    let cnt = 1;

    while (head < q.length) {
      const [x, y] = q[head++];

      for (let k = 0; k < 4; k++) {
        const nx = x + dx[k];
        const ny = y + dy[k];

        if (0 <= nx && nx < rows && 0 <= ny && ny < cols) {
          if (!visited[nx][ny] && arr[nx][ny] === color) {
            visited[nx][ny] = true;
            q.push([nx, ny]);
            cnt++;
          }
        }
      }
    }
    return cnt;
  };

  let w_total = 0;
  let b_total = 0;

  for (let i = 0; i < rows; i++) {
    for (let j = 0; j < cols; j++) {
      if (!visited[i][j]) {
        const color = arr[i][j];
        const size = bfs(i, j, color);
        if (color === "W") w_total += size * size;
        else b_total += size * size;
      }
    }
  }

  console.log(w_total, b_total);
};

solution(input);
